File: backend/src/services/s3EmailService.js
/**
 * S3 Email Service
 * Downloads and manages emails stored in S3 by AWS SES
 */

const { S3Client, GetObjectCommand, DeleteObjectCommand } = require('@aws-sdk/client-s3');
const { query } = require('../config/database');

// Initialize S3 client
const s3Client = new S3Client({
  region: process.env.AWS_REGION || 'us-east-1',
  credentials: process.env.AWS_ACCESS_KEY_ID ? {
    accessKeyId: process.env.AWS_ACCESS_KEY_ID,
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY
  } : undefined // Use IAM role if no explicit credentials
});

const BUCKET_NAME = process.env.INBOUND_EMAIL_BUCKET || 'isrs-inbound-emails';

/**
 * Download email from S3
 * @param {string} s3Key - S3 object key
 * @returns {Promise<string>} Raw email content
 */
async function downloadEmailFromS3(s3Key) {
  try {
    const command = new GetObjectCommand({
      Bucket: BUCKET_NAME,
      Key: s3Key
    });

    const response = await s3Client.send(command);

    // Convert stream to string
    const chunks = [];
    for await (const chunk of response.Body) {
      chunks.push(chunk);
    }
    const rawEmail = Buffer.concat(chunks).toString('utf-8');

    console.log(`[S3 Email] Downloaded email from S3: ${s3Key} (${rawEmail.length} bytes)`);

    return rawEmail;
  } catch (error) {
    console.error(`[S3 Email] Error downloading from S3:`, error);
    throw new Error(`Failed to download email from S3: ${error.message}`);
  }
}

/**
 * Delete email from S3 after processing
 * @param {string} s3Key - S3 object key
 */
async function deleteEmailFromS3(s3Key) {
  try {
    const command = new DeleteObjectCommand({
      Bucket: BUCKET_NAME,
      Key: s3Key
    });

    await s3Client.send(command);
    console.log(`[S3 Email] Deleted email from S3: ${s3Key}`);
  } catch (error) {
    console.error(`[S3 Email] Error deleting from S3:`, error);
    // Don't throw - deletion is not critical (lifecycle rule will clean up)
  }
}

/**
 * Check if email has already been processed (duplicate detection)
 * @param {string} messageId - Email message ID
 * @returns {Promise<boolean>} True if already processed
 */
async function isEmailProcessed(messageId) {
  const result = await query(
    'SELECT id FROM parsed_emails WHERE metadata->>\'message_id\' = $1',
    [messageId]
  );
  return result.rows.length > 0;
}

/**
 * Mark email as processed
 * @param {string} s3Key - S3 object key
 * @param {string} messageId - Email message ID
 * @param {number} parsedEmailId - Parsed email database ID
 */
async function markEmailAsProcessed(s3Key, messageId, parsedEmailId) {
  await query(
    `INSERT INTO processed_s3_emails (s3_key, message_id, parsed_email_id, processed_at)
     VALUES ($1, $2, $3, NOW())
     ON CONFLICT (s3_key) DO NOTHING`,
    [s3Key, messageId, parsedEmailId]
  );
}

/**
 * Get S3 bucket configuration
 * @returns {Object} Bucket configuration
 */
function getS3Config() {
  return {
    bucket: BUCKET_NAME,
    region: process.env.AWS_REGION || 'us-east-1',
    hasCredentials: !!process.env.AWS_ACCESS_KEY_ID
  };
}

module.exports = {
  downloadEmailFromS3,
  deleteEmailFromS3,
  isEmailProcessed,
  markEmailAsProcessed,
  getS3Config
};

File: backend-python/app/services/s3_email_service.py
"""
S3 Email Service
Downloads and manages emails from S3 bucket
"""
import logging
from typing import Optional
import boto3
from botocore.exceptions import ClientError
from app.config import settings

logger = logging.getLogger(__name__)


class S3EmailService:
    """Service for downloading emails from S3"""

    def __init__(self):
        """Initialize S3 client"""
        self.s3_client = boto3.client(
            's3',
            region_name=settings.AWS_REGION,
            aws_access_key_id=settings.AWS_ACCESS_KEY_ID,
            aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY
        )
        self.bucket_name = settings.INBOUND_EMAIL_BUCKET

    async def download_email(self, s3_key: str) -> Optional[bytes]:
        """
        Download email content from S3

        Args:
            s3_key: S3 object key for the email

        Returns:
            Email content as bytes, or None if download fails
        """
        try:
            logger.info(f"[S3] Downloading email from s3://{self.bucket_name}/{s3_key}")

            response = self.s3_client.get_object(
                Bucket=self.bucket_name,
                Key=s3_key
            )

            email_content = response['Body'].read()
            logger.info(f"[S3] Successfully downloaded email ({len(email_content)} bytes)")

            return email_content

        except ClientError as e:
            error_code = e.response.get('Error', {}).get('Code', 'Unknown')
            logger.error(f"[S3] Failed to download email: {error_code} - {str(e)}")
            return None

        except Exception as e:
            logger.error(f"[S3] Unexpected error downloading email: {str(e)}", exc_info=True)
            return None

    async def check_email_exists(self, s3_key: str) -> bool:
        """
        Check if an email exists in S3

        Args:
            s3_key: S3 object key for the email

        Returns:
            True if email exists, False otherwise
        """
        try:
            self.s3_client.head_object(
                Bucket=self.bucket_name,
                Key=s3_key
            )
            return True

        except ClientError as e:
            error_code = e.response.get('Error', {}).get('Code', 'Unknown')
            if error_code == '404':
                return False
            logger.error(f"[S3] Error checking email existence: {error_code} - {str(e)}")
            return False

        except Exception as e:
            logger.error(f"[S3] Unexpected error checking email existence: {str(e)}")
            return False
